### ISSUE TITLE
get returns random values when lower than default OPT_POLL_TIMEOUT is set

### ISSUE THREAD

**Issue by ondrejcech** [php-memcached-dev/php-memcached#531]
After upgrading to libmemcached to 3.1.5-6, I'm sometimes (!) getting random results when calling the get method. There's no error, but the returned value is from a different key. This does not happen on every request. I've narrowed this error down to `OPT_POLL_TIMEOUT` option. When set to lower than 500ms this starts to happen. It seems that if `OPT_POLL_TIMEOUT` happens, the library returns a seemingly random value without any error. 

It does not happen when `OPT_POLL_TIMEOUT` is set to default 1000ms value or omitted.

Not sure if it's an issue of libmemcached or php-memcached though. 

PHP version:
```
php -v
PHP 8.1.12 (cli) (built: Oct 25 2022 17:30:00) (NTS gcc x86_64)
Copyright (c) The PHP Group
Zend Engine v4.1.12, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.12, Copyright (c), by Zend Technologies
```

memcache* libraries:
```
rpm -qa | grep memcached
remi-libmemcached-awesome-1.1.3-1.el7.remi.x86_64
libmemcached-1.0.16-5.el7.x86_64
php81-php-pecl-memcached-3.2.0-1.el7.remi.x86_64
libmemcached-opt-libs-1.0.18-2.el7.remi.x86_64
memcached-1.6.16-1.el7.remi.x86_64
php80-php-pecl-memcached-3.1.5-4.el7.remi.x86_64
```

some memcache stats
```
echo 'stats' | nc localhost 11211
STAT version 1.6.16
STAT libevent 2.0.21-stable
END
```

I will be more than happy to provide more debug info.

**Comment by ondrejcech** [comment]
@sodabrew TBH, I think most of the thumbs-up comes from the same project. That volume just means how important the issue is for us.

**Comment by sodabrew** [comment]
This is awesome debugging info! Tag @m6w6, and maybe just open a tracking issue on awesomized/libmemcahed that links back here.

**Comment by m6w6** [comment]
Thanks for your effort!

Any chance for some code that would reproduce this issue?

It's unfortunate, that it has always been tried to upgrade both at the same time, so neither pecl-memcached nor libmemcached-awesome can be ruled out...

That's suspicious, at least:
> (It looks like) it is only reproduced if persistent_id is set during Memcached construction and is not unique.


I'll have a look at `POLL_TIMEOUT` in the meantime, but TBH, it doesn't look too promising that I'll "just be able to see" something...

**Comment by k0ka** [comment]
I can confirm setting `OPT_POLL_TIMEOUT` to default value `1000` fixes the problem.

I guess it's some sort of "race condition" bug. I tried to run a simple script with a lot of gets, but the bug wasn't reproduced on it. On my test server it is randomly shown once per 5-10 requests.

**Comment by m6w6** [comment]
It could be, that I found a [related commit](https://github.com/awesomized/libmemcached/commit/011ea033d902722b9744ea9f28e646734ffdcaf2). Hang on.

**Comment by k0ka** [comment]
I'm still debugging it. I already found that it works perfectly until timeout occurs. 

it returns an error like this
`(0x5585c02958b0) A TIMEOUT OCCURRED, time out,  host: localhost:11211 -> /builddir/build/BUILD/libmemcached-6e713c39031ff312164cc3ff12f943f8e8e01885/src/libmemcached/io.cc:220`

And after that it starts sending wrong responses quite often until process is restarted. 
I'm going to continue debugging it soon using gdb and most probably will post here some results tomorrow.

**Comment by k0ka** [comment]
I guess there are two bad things:
1) `libmemcached-awesome` sometimes throws `A TIMEOUT OCCURRED` even though it must not happen. 20ms poll timeout is more than enough for a localhost connection. 
2) `php-pecl-memcached` does something wrong when this timeout comes in. 

I added logs to php to print everything that is returned from `Memcached::get` immediately after it. Here is the log of one failed page request:

```
    [0] => Array
        (
            [type] => get
            [key] => joyreactor:symfony.routing.data
            [value] => a:104:{s:52:"parse_/post/4849934_1321ec5d5af53bbf347fa27ea57d3d4b";...
            [error] => SUCCESS
        )

    [1] => Array
        (
            [type] => get
            [key] => joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__postSeo/sf_cache_key/4849934_v6
            [value] => a:2:{s:7:"content";s:514:"<div class="sidebar_block"><h2 class="sideheader">...
            [error] => SUCCESS
        )

    [2] => Array
        (
            [type] => get
            [key] => joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__postSeo/sf_cache_key/4849934_v6
            [value] => a:2:{s:7:"content";s:1742:"0|{"title":"\u041a\u0430\u043c\u0435\u043d\u043d\u043e\u0443\u0433\u043e...                  
            [error] => SUCCESS
        )

    [3] => Array
        (
            [type] => get
            [key] => joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__post_cached/sf_cache_key/4849934_noauth
            [value] => a:2:{s:7:"content";s:2512:"0|<div class="article post-normal"><div class="post_top">...
            [error] => SUCCESS
        )
```


We have 4 requests there. The 2nd and 3rd request requests the same key. The return value on the 2nd request is not right. It returned a value from another key (that was in the end of the previous request). The right value is returned in the 3rd response. 

I set 2 breakpoints in the `libmemcached-awesome` and returned key that was queried there:
1)
 ```
br memcached_mget
printf "mget = %s\n", keys[0]
```
2) 
```
br memcached_fetch_result
printf "fetch = %s\n", result->item_key
```

Here is what was printed:

```
mget = joyreactor:routing:4:symfony.routing.data
fetch =
fetch = joyreactor:routing:4:symfony.routing.data▒o%
mget = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__postSeo/sf_cache_key/4849934_v6
fetch =
fetch = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/people/__onlineUsers/sf_cache_key/40cd750bba9870f18aada2478b24840a
fetch = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__postSeo/sf_cache_key/4849934_v6750bba9870f18aada2478b24840a
mget = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__postSeo/sf_cache_key/4849934_v6
fetch =
fetch = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__postSeo/sf_cache_key/4849934_v6
mget = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__post_cached/sf_cache_key/4849934_noauth
fetch =
fetch = joyreactor:/koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/post/__post_cached/sf_cache_key/4849934_noauth
```

As you can see, after the 2nd mget it requests old key first (`__onlineUsers`) and then requests the right key (`__postSeo`) but ignores its result.

In both logs I skipped everything after the 4th request. There was 19 requests to memcached on the page.

**Comment by k0ka** [comment]
Let me know if you need any other internal information to debug it.

**Comment by m6w6** [comment]
This might be an even better candidate: https://github.com/awesomized/libmemcached/commit/d7a0084bf99d618d1dc26a54fd413db7ae8b8e63

**Comment by m6w6** [comment]
I'd need someone to confirm/decline any offending commit to proceed; or/and (even better) a reliable reproduce case.

**Comment by m6w6** [comment]
Hey! Did anyone have a chance to test whether their issue was solved with reverting either commit?

**Comment by k0ka** [comment]
Hi, 

Sorry for long response. I set the test environment. Reverted that commit (https://github.com/k0ka/libmemcached/commit/61976509ab214ee5f73b4a23de3eaa727cd2cd43). But when I'm building rpm with this patch (from Remi's srpm) it hangs on test 31: `memcached_noblock`.

Should I try to revert only changes in `response.cc`?

**Comment by m6w6** [comment]
I remember fiddling around with this test when looking for offending commits. It's probably a bad test in itself.

IIRC, just make https://github.com/k0ka/libmemcached/blob/v1.x/test/tests/memcached/noblock.cpp#L13 read:
```cpp
auto num = 10000;
```

**Comment by m6w6** [comment]
That's interesting. It doesn't hang, though; it just takes a very long time on linux.
See https://github.com/awesomized/libmemcached/actions/runs/4301629917

**Comment by m6w6** [comment]
Thanks, for the test case, though. I'll try to debug it on Linux later.

**Comment by k0ka** [comment]
I also don't understand why it returns `MEMCACHED_NOTFOUND` sometimes. 
The main idea of the test if just to set 500 keys and then query them several times.

**Comment by k0ka** [comment]
In your case `timeoutAchieved` is not set to true (if it is set the test will fail). On my test server it happens every run. Maybe you should decrease `MEMCACHED_BEHAVIOR_POLL_TIMEOUT` ? 

I tested it with MEMCACHED_BEHAVIOR_POLL_TIMEOUT=1000 and 20. It never happens on 1000 and always happens on 20.

**Comment by k0ka** [comment]
So, I'm using `rockylinux 8` (RHEL clone). Compile flags are taken from the remi's srpm:

`
/usr/bin/cmake -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=/opt/libmemcached-awesome -DINCLUDE_INSTALL_DIR:PATH=/opt/libmemcached-awesome/include -DLIB_INSTALL_DIR:PATH=/opt/libmemcached-awesome/lib64 -DSYSCONF_INSTALL_DIR:PATH=/etc -DSHARE_INSTALL_PREFIX:PATH=/opt/libmemcached-awesome/share -DLIB_SUFFIX=64 -DBUILD_SHARED_LIBS:BOOL=ON -DBUILD_TESTING:BOOL=ON -DBUILD_DOCS_MAN:BOOL=ON -DBUILD_DOCS_MANGZ:BOOL=OFF -DENABLE_SASL:BOOL=ON -DENABLE_DTRACE:BOOL=ON -DENABLE_OPENSSL_CRYPTO:BOOL=ON -DENABLE_HASH_HSIEH:BOOL=ON -DENABLE_HASH_FNV64:BOOL=ON -DENABLE_HASH_MURMUR:BOOL=ON -DENABLE_MEMASLAP:BOOL=ON -S .
`

**Comment by m6w6** [comment]
Thanks for the extra info. Can you also share some words about your hardware?

**Comment by k0ka** [comment]
```
[root@serv52 koka]# lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              12
On-line CPU(s) list: 0-11
Thread(s) per core:  2
Core(s) per socket:  6
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
BIOS Vendor ID:      Advanced Micro Devices, Inc.
CPU family:          23
Model:               113
Model name:          AMD Ryzen 5 3600 6-Core Processor
BIOS Model name:     AMD Ryzen 5 3600 6-Core Processor
Stepping:            0
CPU MHz:             3600.000
CPU max MHz:         4208.2031
CPU min MHz:         2200.0000
BogoMIPS:            7185.82
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-11
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
[root@serv52 koka]# uname -a
Linux serv52.identityworks.ru 4.18.0-372.9.1.el8.x86_64 #1 SMP Tue May 10 14:48:47 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by k0ka** [comment]
I finally created a docker container with this patch. It looks like this bug is gone - I wasn't able to reproduce it with that commit reverted. However I still see quite a lot of "TIMEOUT OCCURRED" errors. About 5%. The memcached is on the same server. 

```
    [14] => Array
        (
            [type] => get
            [key] => /koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/people/__onlineUsers/sf_cache_key/40cd750bba9870f18aada2478b24840a
            [value] => 
            [error] => (0x55fa3332e5d0) A TIMEOUT OCCURRED, time out,  host: localhost:11211 -> /root/rpmbuild/BUILD/libmemcached-6e713c39031ff312164cc3ff12f943f8e8e01885/src/libmemcached/io.cc:220
        )

    [15] => Array
        (
            [type] => get
            [key] => /koka_joyreactor_identityworks_ru/ru/all/sf_cache_partial/people/__onlineUsers/sf_cache_key/40cd750bba9870f18aada2478b24840a
            [value] => 
            [error] => SUCCESS
        )
```

**Comment by m6w6** [comment]
Just realised, your test is doing `500!` gets on 500 existing and `500! - 500` non-existent keys... (! means factorial)... yikes! :rofl: 

Anyway, I *think*, in normal circumstances, with this low of a timeout, there's still the previous response floating around, since, well, a timeout occurred while trying to fetch it, and the next request will happily fetch the previous response the next time.

I also *think*, these timeouts are really low, even computers sometimes need some time to think :thinking: 
If the server is going to fail, it will also fail with a timeout of 100ms.

Please have a try with awesomized/libmemcached@48dcc61a and see if it fixes your issues; it basically reverts 90% of awesomized/libmemcached@d7a0084bf99d618d1dc26a54fd413db7ae8b8e63

Thank you!

**Comment by k0ka** [comment]
it's not 500! iterations, it's only 500*250 iterations. 

And 100ms is way too much for a bare metal server. The full page is generated in 200-300ms including several sql requests and 15-20 memcached requests. It took less than 1ms to process a memcached request - you may check my tcpdump log. It has timestamps.

**Comment by m6w6** [comment]
> it's not 500! iterations, it's only 500*250 iterations. 

Uhm, I may have been sidetracked, please enlighten me!


> And 100ms is way too much for a bare metal server. The full page is generated in 200-300ms including several sql requests and 15-20 memcached requests. It took less than 1ms to process a memcached request - you may check my tcpdump log. It has timestamps. 

I unterstand your sentiment, but i have been trying to understand Linux clock granularity for the past couple hours.

If a server fails, it doesn't matter 
If in 20 or 200 ms.

Most probably I've been looking in the wrong corners...

**Comment by k0ka** [comment]
> Please have a try with [awesomized/libmemcached@48dcc61a](https://github.com/awesomized/libmemcached/commit/48dcc61a) and see if it fixes your issues; it basically reverts 90% of [awesomized/libmemcached@d7a0084](https://github.com/awesomized/libmemcached/commit/d7a0084bf99d618d1dc26a54fd413db7ae8b8e63)

I checked it, it works the same way - it fixes the most nasty bug with "wrong key return". However the timeout still occurs. 

I'm pretty sure it's not right. I can get tcpdump of interaction with memcached server to see timings.

**Comment by k0ka** [comment]
> > it's not 500! iterations, it's only 500*250 iterations.
> 
> Uhm, I may have been sidetracked, please enlighten me!

The current test is:
```
1) create 500 keys.
2) for i from 0 to 500:
        for j from 0 to i:
            get j-th key
```

so the first iteration goes from 0 to 0 and asks for 1 keys, the 2nd - from 0 to 1 and asks for 2 keys, 3rd - from 0 to 2 and asks 3 keys, nth goes from 0 to n and asks for n keys.
1 + 2 + 3 + ... + n = (n * n-1)/2

(actually it goes from 0 to i-1, so it asks even less keys)

**Comment by m6w6** [comment]
Yep, but for ` num` times.

**Comment by k0ka** [comment]
why num times? it executes `memcached_mget` with 1 key inside of two cycles. And fetches only 1 key. 

Anyway, I narrowed my test to this:
```C++
#include "test/lib/common.hpp"
#include "test/lib/MemcachedCluster.hpp"

TEST_CASE("memcached_poll_timeout") {
  auto test = MemcachedCluster{Cluster{Server{MEMCACHED_BINARY, {"-p", "11211"}}, 1}};
  auto memc = &test.memc;

  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_NO_BLOCK, 1));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL, 0));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_TCP_NODELAY, 0));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_POLL_TIMEOUT, 20));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_RCV_TIMEOUT, 20 * 1000));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_SND_TIMEOUT, 20 * 1000));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_RETRY_TIMEOUT, 60));
  // REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_CONNECT_TIMEOUT, 20));

  memcached_return_t rc;

  auto num = 100'000;

  REQUIRE_SUCCESS(memcached_flush(memc, 0));

  const char* key = "testkey1";
  auto len = strlen(key);

  REQUIRE_SUCCESS(memcached_set(memc, key, len, key, len, 1, 0));
  memcached_quit(memc);

  const char *keys[1] = {key};

  for (auto i = 0; i < num; ++i) {
    REQUIRE_SUCCESS(memcached_mget(memc, keys, &len, 1));

    auto numResults = 0;
    while (auto rv = memcached_fetch(memc, nullptr, nullptr, nullptr, nullptr, &rc)) {
      REQUIRE_SUCCESS(rc);
      free(rv);

      numResults++;
      REQUIRE(numResults == 1);
    }

    if (MEMCACHED_END == rc) {
      continue;
    }

    INFO(rc << " on iteration " << i);
    REQUIRE(false);
  }
}
```

So it is only set one key `testkey1` and mget's it 100'000 times. 

It constantly fails for me on about 60-70 000 iteration. But it is the `memcached` server fail.

```
19:33:10.916117 IP (tos 0x0, ttl 64, id 45519, offset 0, flags [DF], proto TCP (6), length 66)
    127.0.0.1.56098 > 127.0.0.1.11211: Flags [P.], cksum 0xfe36 (incorrect -> 0xec93), seq 978181:978195, ack 2445451, win 128, options [nop,nop,TS val 2401858559 ecr 2401858559], length 14
        0x0000:  4500 0042 b1cf 4000 4006 8ae4 7f00 0001  E..B..@.@.......
        0x0010:  7f00 0001 db22 2bcb e142 3258 8ddf 27d6  ....."+..B2X..'.
        0x0020:  8018 0080 fe36 0000 0101 080a 8f29 73ff  .....6.......)s.
        0x0030:  8f29 73ff 6765 7420 7465 7374 6b65 7931  .)s.get.testkey1
        0x0040:  0d0a                                     ..
19:33:10.916121 IP (tos 0x0, ttl 64, id 56988, offset 0, flags [DF], proto TCP (6), length 87)
    127.0.0.1.11211 > 127.0.0.1.56098: Flags [P.], cksum 0xfe4b (incorrect -> 0x3301), seq 2445451:2445486, ack 978195, win 128, options [nop,nop,TS val 2401858559 ecr 2401858559], length 35
        0x0000:  4500 0057 de9c 4000 4006 5e02 7f00 0001  E..W..@.@.^.....
        0x0010:  7f00 0001 2bcb db22 8ddf 27d6 e142 3266  ....+.."..'..B2f
        0x0020:  8018 0080 fe4b 0000 0101 080a 8f29 73ff  .....K.......)s.
        0x0030:  8f29 73ff 5641 4c55 4520 7465 7374 6b65  .)s.VALUE.testke
        0x0040:  7931 2030 2038 0d0a 7465 7374 6b65 7931  y1.0.8..testkey1
        0x0050:  0d0a 454e 440d 0a                        ..END..
19:33:10.916129 IP (tos 0x0, ttl 64, id 45520, offset 0, flags [DF], proto TCP (6), length 66)
    127.0.0.1.56098 > 127.0.0.1.11211: Flags [P.], cksum 0xfe36 (incorrect -> 0xec62), seq 978195:978209, ack 2445486, win 128, options [nop,nop,TS val 2401858559 ecr 2401858559], length 14
        0x0000:  4500 0042 b1d0 4000 4006 8ae3 7f00 0001  E..B..@.@.......
        0x0010:  7f00 0001 db22 2bcb e142 3266 8ddf 27f9  ....."+..B2f..'.
        0x0020:  8018 0080 fe36 0000 0101 080a 8f29 73ff  .....6.......)s.
        0x0030:  8f29 73ff 6765 7420 7465 7374 6b65 7931  .)s.get.testkey1
        0x0040:  0d0a                                     ..
19:33:10.916149 IP (tos 0x0, ttl 64, id 56989, offset 0, flags [DF], proto TCP (6), length 57)
    127.0.0.1.11211 > 127.0.0.1.56098: Flags [P.], cksum 0xfe2d (incorrect -> 0x0e02), seq 2445486:2445491, ack 978209, win 128, options [nop,nop,TS val 2401858560 ecr 2401858559], length 5
        0x0000:  4500 0039 de9d 4000 4006 5e1f 7f00 0001  E..9..@.@.^.....
        0x0010:  7f00 0001 2bcb db22 8ddf 27f9 e142 3274  ....+.."..'..B2t
        0x0020:  8018 0080 fe2d 0000 0101 080a 8f29 7400  .....-.......)t.
        0x0030:  8f29 73ff 454e 440d 0a                   .)s.END..
```

I guess we might accept https://github.com/awesomized/libmemcached/commit/48dcc61a as the fix for this problem.

**Comment by k0ka** [comment]
wait, I'm dumb. I set key to 1 sec lifetime. That's why it is expired. Making more testing...

**Comment by k0ka** [comment]
Or I'm a genius (accidentally)! It looks like problems starts when a key is expired. It only happens when `BINARY_PROTOCOL` is on. The problem is in the memcached server. It just answers very long for a key that doesn't exist. 


```C++
#include "test/lib/common.hpp"
#include "test/lib/MemcachedCluster.hpp"

TEST_CASE("memcached_poll_timeout") {
  auto test = MemcachedCluster{Cluster{Server{MEMCACHED_BINARY, {"-p", "11211"}}, 1}};
  auto memc = &test.memc;

  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_NO_BLOCK, 1));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_TCP_NODELAY, 0));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL, 1));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_POLL_TIMEOUT, 20));

  memcached_return_t rc;

  REQUIRE_SUCCESS(memcached_flush(memc, 0));

  const char* key = "testkey1";
  auto len = strlen(key);

  const char *keys[1] = {key};

  REQUIRE_SUCCESS(memcached_mget(memc, keys, &len, 1));

  auto numResults = 0;
  while (auto rv = memcached_fetch(memc, nullptr, nullptr, nullptr, nullptr, &rc)) {
    REQUIRE_SUCCESS(rc);
    free(rv);

    numResults++;
    REQUIRE(numResults == 1);
  }

  if (MEMCACHED_END != rc && MEMCACHED_NOTFOUND != rc) {
    INFO(rc);
    REQUIRE(false);
  }
}
```

This test always fails with rc=31 (MEMCACHED_TIMEOUT) for me. Setting `BINARY_PROTOCOL` to `0` makes the test passes perfectly. If I set `POLL_TIMEOUT` to `100` the test will pass but fetch would take quite a lot of time (about 40ms for each request).

TL;DR:
1) `BINARY_PROTOCOL` must be set to 0 in production (I heard about it somewhere already). 
2) querying non-existent key with `BINARY_PROTOCOL`=1 and small `POLL_TIMEOUT` is a good way to reliably get `MEMCACHED_TIMEOUT` error for tests to simulate a server that is not completely down.
3) It looks like https://github.com/awesomized/libmemcached/commit/48dcc61a fixes the problem.

**Comment by k0ka** [comment]
And here is a test that does it:

```C++
#include "test/lib/common.hpp"
#include "test/lib/MemcachedCluster.hpp"

TEST_CASE("memcached_poll_timeout") {
  auto test = MemcachedCluster{Cluster{Server{MEMCACHED_BINARY, {"-p", "11211"}}, 1}};
  auto memc = &test.memc;

  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_NO_BLOCK, 1));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_TCP_NODELAY, 0));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_BINARY_PROTOCOL, 1));
  REQUIRE_SUCCESS(memcached_behavior_set(memc, MEMCACHED_BEHAVIOR_POLL_TIMEOUT, 20));

  memcached_return_t rc;
  size_t len;

  auto num = 100;
  auto hit = 0u;
  auto rnd = "test_key_";

  REQUIRE_SUCCESS(memcached_flush(memc, 0));

  for (auto i = 1; i < num; ++i) {
    auto setKey = rnd + to_string(i);
    REQUIRE_SUCCESS(memcached_set(memc, setKey.c_str(), setKey.length(), setKey.c_str(),
                                  setKey.length(), 0, 0));
  }
  memcached_quit(memc);

  auto timeoutAchieved = false;
  for (auto i = 0; i < num; ++i) {
    for (auto j = 0; j <= i; ++j) {
      auto getKey = rnd + to_string(j);
      auto len = getKey.length();
      char key[MEMCACHED_MAX_KEY];
      memcpy(key, getKey.c_str(), len + 1);
      const char *keys[1] = {key};

      REQUIRE_SUCCESS(memcached_mget(memc, keys, &len, 1));

      auto numResults = 0;
      while (auto rv = memcached_fetch(memc, key, nullptr, &len, nullptr, &rc)) {
        REQUIRE_SUCCESS(rc);
        free(rv);

        numResults++;
        REQUIRE(numResults == 1);
      }

      if (MEMCACHED_END == rc || MEMCACHED_NOTFOUND == rc ){
        continue;
      }

      if (MEMCACHED_TIMEOUT == rc){
        timeoutAchieved = true;
        continue;
      }

      INFO(rc);
      REQUIRE(false);
    }
  }

  REQUIRE(timeoutAchieved);
}
```

On my server it always fails on the current version of libmemcached, but always passes after https://github.com/awesomized/libmemcached/commit/48dcc61a is applied.

Here is what it does:
1) it creates keys from 1 to num. It doesn't create key 0.
2) for i from 0 to num-1, it asks for keys 0, 1, 2, ... i, i in a separate mget requests. Each mget requests only one key. 
3) the key 0 doesn't exist so memcached most probably would return `MEMCACHED_TIMEOUT`.
4) we are querying other (i-1) keys after timeout. And check that only one result is returned. 
5) at some point it returns two results via `memcached_fetch` and assertion `REQUIRE(numResults == 1);` fails

**Comment by m6w6** [comment]
Hey, I pushed a much simpler test case; in case you didn't notice.

Unfortunately the fix for this problem is simple & hard (like in the branch) or complicated & considerate (I'll try to write something up until tomorrow).

If your observation of "memcached is choking on missing-key+binary-proto", we should file a decent issue on https://github.com/memcached/memcached, IMO. I'll also try to have a look at this.

Finally, I'm definitely *not* understanding why `memcached_io_poll` immediately returns when the timeout is set to zero? Maybe the socket actually IS readable without a timeout?

Unfortunately, the libmemcached legacy is a over-engineered mess; someone should think about making a new meta-protocol-only client library. Maybe @dormando's proxy code would be a good fit for a new client library?

**Comment by dormando** [comment]
Hi. Sorry I've only skimmed this; that sounds like a nasty bug.

RE client replacement, this issue on php-memcached is probably the wrong place to discuss, so I created an issue topic over here: https://github.com/dormando/mcmc/issues/2

TL:DR; I wrote a core of a client parser (it's messy because I didn't know what direction to go in at the time) which is in use for the proxy code + some other utilities. I'm trying to find time to expand this into a full client soon but am not sure what a useful replacement looks like.

I wrote up my goals and would invite anyone on this thread who cares (certainly @m6w6 as you seem to have taken ownership of libmemcached) to come help with at least feedback if not code.

**Comment by WPprodigy** [comment]
Wanted to mention (for others that may come looking) that I've seen this occurring even without `OPT_POLL_TIMEOUT` configured. The only two possibly related options that were set are:

```
Memcached::OPT_CONNECT_TIMEOUT => 1000,
Memcached::OPT_TCP_NODELAY     => true,
```

Notably this issue only occurs when the memcached server nodes are experiencing some sort of stress, like kernel blocking under memory pressure and page eviction.

I've had a hard time replicating without `OPT_POLL_TIMEOUT`, but can do so consistently with that option here https://github.com/WPprodigy/memcached-inconsistency-debugging/ by sending pause/continue signals to the memcached servers . And can also confirm that updating libmemcached from 1.1.3 to 1.1.4 does resolve what I've replicated in that test environment.

**Comment by ondrejcech** [comment]
Because we've moved to Redis, I can't tell if this issue is fixed by ibmemcached 1.1.4. Can someone else please confirm that so I can close this issue? cc @k0ka @m6w6

**Comment by k0ka** [comment]
yep, it was fixed

**Comment by ondrejcech** [comment]
Thank you!

**Comment by sodabrew** [comment]
Thank you all for the report, reproduction case, and follow-up that this was fixed upstream!
